x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java

```diff
@@ -291,59 +308,116 @@ static Map<FullyQualifiedInferenceId, InferenceResults> getInferenceResults(
+                    Map.of(),
+                    InputType.INTERNAL_SEARCH,
+                    null,
+                    false
+                )
             )
-        );
+            .toList();
+
+        queryRewriteContext.registerAsyncAction((client, listener) -> {
+            GroupedActionListener<Tuple<FullyQualifiedInferenceId, InferenceResults>> gal = createGroupedActionListener(
+                inferenceResultsMapSupplier,
+                inferenceRequests.size(),
+                listener
+            );
+            for (InferenceAction.Request inferenceRequest : inferenceRequests) {
+                FullyQualifiedInferenceId fullyQualifiedInferenceId = new FullyQualifiedInferenceId(
+                    queryRewriteContext.getLocalClusterAlias(),
+                    inferenceRequest.getInferenceEntityId()
+                );
+                executeAsyncWithOrigin(
+                    client,
+                    ML_ORIGIN,
+                    InferenceAction.INSTANCE,
+                    inferenceRequest,
+                    gal.delegateFailureAndWrap((l, inferenceResponse) -> {
+                        InferenceResults inferenceResults = validateAndConvertInferenceResults(
+                            inferenceResponse.getResults(),
+                            fullyQualifiedInferenceId.inferenceId()
+                        );
+                        l.onResponse(Tuple.tuple(fullyQualifiedInferenceId, inferenceResults));
+                    })
+                );
+            }
+        });
+    }
+
+    static <T extends QueryBuilder> T getNewInferenceResultsFromSupplier(
+        Supplier<Map<FullyQualifiedInferenceId, InferenceResults>> supplier,
+        T currentQueryBuilder,
+        Map<FullyQualifiedInferenceId, InferenceResults> currentInferenceResultsMap,
+        Function<Map<FullyQualifiedInferenceId, InferenceResults>, T> copyGenerator
+    ) {
+        T rewritten = currentQueryBuilder;
+        Map<FullyQualifiedInferenceId, InferenceResults> newInferenceResultsMap = supplier.get();
+        if (newInferenceResultsMap != null) {
+            Map<FullyQualifiedInferenceId, InferenceResults> mergedInferenceResultsMap = mergeInferenceResultsMaps(
+                currentInferenceResultsMap,
+                newInferenceResultsMap
+            );
+            rewritten = copyGenerator.apply(mergedInferenceResultsMap);
+        }
+
+        return rewritten;
+    }
+
+    private static GroupedActionListener<Tuple<FullyQualifiedInferenceId, InferenceResults>> createGroupedActionListener(
+        InferenceResultsMapSupplier inferenceResultsMapSupplier,
+        int inferenceRequestCount,
+        ActionListener<?> listener
+    ) {
+        return new GroupedActionListener<>(inferenceRequestCount, listener.delegateFailureAndWrap((l, responses) -> {
+            ImmutableOpenMap.Builder<FullyQualifiedInferenceId, InferenceResults> mapBuilder = ImmutableOpenMap.builder();
```

Comment:
I don't understand why we are using ImmutableOpenMap. Just use a hashmap with a preallocated size.